ttnn/cpp/ttnn/tensor/types.hpp

```diff
@@ -514,42 +517,46 @@ struct MultiDeviceHostStorage {
         mutable std::mutex shape_mtx;
         MultiDeviceStorage() = default;
 
+        friend void swap(MultiDeviceStorage& first, MultiDeviceStorage& second) {
+            std::scoped_lock lock(first.buffer_mtx, first.shape_mtx, second.buffer_mtx, second.shape_mtx);
+            // enable ADL (not necessary, but good practice)
+            using std::swap;
+
+            swap(first.strategy, second.strategy);
+            swap(first.ordered_device_ids, second.ordered_device_ids);
+            swap(first.buffers, second.buffers);
+            swap(first.shapes, second.shapes);
+        }
+
         MultiDeviceStorage(
             DistributedTensorConfig strategy_,
             std::vector<int> ordered_device_ids_,
             std::unordered_map<int, DeviceBuffer> buffers_,
-            std::unordered_map<int, Shape> shapes_) : strategy(strategy_), ordered_device_ids(ordered_device_ids_), buffers(buffers_), shapes(shapes_) {}
+            std::unordered_map<int, Shape> shapes_)
+                : strategy(std::move(strategy_)),
+                ordered_device_ids(std::move(ordered_device_ids_)),
+                buffers(std::move(buffers_)), shapes(std::move(shapes_)) {}
 
         MultiDeviceStorage(MultiDeviceStorage &&other) {
-            std::scoped_lock buf_lock(buffer_mtx, shape_mtx);
-            ordered_device_ids = other.ordered_device_ids;
-            strategy = other.strategy;
-            buffers = other.buffers;
-            shapes = other.shapes;
+            swap(*this, other);
         }
+
         MultiDeviceStorage(const MultiDeviceStorage &other) {
-            std::scoped_lock buf_lock(buffer_mtx, shape_mtx);
+            std::scoped_lock lock(buffer_mtx, shape_mtx, other.buffer_mtx, other.shape_mtx);
```

Comment:
I believe that we can lock only two mutexes from `other`. `buffer_mtx` and `shape_mtx` won't be reachable before this method finish it's execution.

Reply:
yeah, this seems like undefined behavior? Shouldnt need to lock before it's constructed?

Reply:
Mutexes were already created. It is not UB but yeah we can lock only others.